tsdb/engine/tsm1/timestamp.go

```diff
@@ -0,0 +1,309 @@
+//
+// Each compressed byte slice has a 1 byte header indicating the compression type.  The 4 high bits
+// indicated the encoding type.  The 4 low bits are using by the encoding type.
+//
+// For run length encoding, the 4 low bits store the log10 of the scaling factor.  The next 8 bytes are
+// the starting timestamp, next 1-10 bytes is the delta value using variable-length encoding, finally the
+// next 1-10 bytes is the count of values.
+//
+// For simple8b encoding, the 4 low bits store the log10 of the scaling factor.  The next 8 bytes is the
+// first delta value stored uncompressed, the remaining bytes are 64bit words containg compressed delta
+// values.
+//
+// For uncompressed encoding, the delta values are stored using 8 bytes each.
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/jwilder/encoding/simple8b"
+)
+
+const (
+	// timeUncompressed is a an uncompressed format using 8 bytes per timestamp
+	timeUncompressed = 0
+	// timeCompressedPackedSimple is a bit-packed format using simple8b encoding
+	timeCompressedPackedSimple = 1
+	// timeCompressedRLE is a run-length encoding format
+	timeCompressedRLE = 2
+)
+
+// TimeEncoder encodes time.Time to byte slices.
+type TimeEncoder interface {
+	Write(t time.Time)
+	Bytes() ([]byte, error)
+}
+
+// TimeEncoder decodes byte slices to time.Time values.
+type TimeDecoder interface {
+	Next() bool
+	Read() time.Time
+	Error() error
+}
+
+type encoder struct {
+	ts []uint64
+}
+
+// NewTimeEncoder returns a TimeEncoder
+func NewTimeEncoder() TimeEncoder {
+	return &encoder{}
+}
+
+// Write adds a time.Time to the compressed stream.
+func (e *encoder) Write(t time.Time) {
+	e.ts = append(e.ts, uint64(t.UnixNano()))
+}
+
+func (e *encoder) reduce() (max, divisor uint64, rle bool, deltas []uint64) {
+	// Compute the deltas in place to avoid allocating another slice
+	deltas = e.ts
+	// Starting values for a max and divisor
+	max, divisor = 0, 1e12
+
+	// Indicates whether the the deltas can be run-length encoded
+	rle = true
+
+	// Interate in reverse so we can apply deltas in place
+	for i := len(deltas) - 1; i > 0; i-- {
+
+		// First differential encode the values
+		deltas[i] = deltas[i] - deltas[i-1]
+
+		// We're also need to keep track of the max value and largest common divisor
+		v := deltas[i]
+
+		if v > max {
+			max = v
+		}
+
+		for {
+			// If our value is divisible by 10, break.  Otherwise, try the next smallest divisor.
+			if v%divisor == 0 {
+				break
+			}
+			divisor /= 10
+		}
+
+		// Skip the first value || see if prev = curr.  The deltas can be RLE if the are all equal.
+		rle = i == len(deltas)-1 || rle && (deltas[i+1] == deltas[i])
+	}
+	return
+}
+
+// Bytes returns the encoded bytes of all written times.
+func (e *encoder) Bytes() ([]byte, error) {
+	if len(e.ts) == 0 {
+		return []byte{}, nil
+	}
+
+	// Maximum and largest common divisor.  rle is true if dts (the delta timestamps),
+	// are all the same.
+	max, mod, rle, dts := e.reduce()
```

Comment:
Minor, but why call it `mod` here, but `divisor` in the function signature. Small consistencies like this would help with comprehension. Maybe just `div`?